### DISCUSSION TITLE
📚 Docs: How to handle pnpm@10.x.x auto-installing feature

### DISCUSSION THREAD

**Discussion by jakeleventhal** [vercel/turborepo#9522] [Ideas]
### What is the improvement or update you wish to see?

In `pnpm@10`, support will be added for requiring packages to be installed before running commands. I am not sure how turbo will handle this since it runs concurrent processes.
I currently manually do this by having this step in my turbo.json and require every package to depend on this:
```json
    "//#package-deps": {
      "inputs": [
        "!**/.next/**",
        "!**/dist/**",
        "!**/out/**",
        "!**/node_modules/**",
        ".nvmrc",
        "**/package.json",
        "pnpm-lock.yaml",
        "pnpm-workspace.yaml"
      ],
      "outputLogs": "new-only",
      "outputs": ["pnpm-lock.yaml"]
    },
```
It would be nice if turbo supported this natively with pnpm@10

### Is there any context that might help us understand?

N/A

**Comment by anthonyshew** [comment]
Do you have a link to a doc or description for this behavior?

Notably, per [our Support Policy](https://turbo.build/repo/docs/getting-started/support-policy#package-managers), we strive to do our best with package manager compatibility, but likely won't have Day 1 support for new features. I'm only seeing alpha releases for pnpm@10 so we at least would wait for stability and community validation of the feature.

However, from what you're describing, if pnpm simply crashes itself when dependencies aren't updated, then your Turborepo tasks would do the same.

**Comment by jakeleventhal** [comment]
https://github.com/pnpm/pnpm/releases/tag/v10.0.0-alpha.4
> Added a new setting called verify-deps-before-run for checking the state of dependencies before running scripts https://github.com/pnpm/pnpm/issues/8585. The verify-deps-before-run setting supports the following values:

**Comment by anthonyshew** [comment]
Based on my reading of that, Turborepo won't have to do anything, and your `//#package-deps` task will get pushed down into pnpm for native handling. I could be wrong at that, of course, and we'll have to see what is finally released for the feature when it lands in stable. Will leave this open for now until we learn more!

**Reply by jakeleventhal** [reply]
@anthonyshew How will that work if turbo is running multiple pnpm tasks at once?
For example, assume i run `turbo lint test`

Wouldn't that feature mean that both lint and test will try to install packages first (at the same time)?

**Reply by anthonyshew** [reply]
I'm not sure. Just to see what happens today, I ran `pnpm install` in two terminals at the same time in the same repo and one of them crashed, the other one completed successfully and the repo ran fine after that.

If that were the case in your Turborepo tasks, `turbo` would have exited since a task failed. You'd have to be using `--continue` to keep `turbo` running through that failure.

It's hard to say what we'll do in Turborepo until the feature lands.

**Comment by anthonyshew** [comment]
I'm also going to convert this to a Discussion since this is likely may or may not end up being closer to a feature request than a docs update.
